metrics/src/macros.rs

```diff
@@ -0,0 +1,870 @@
+    }
+}
+
+/// Registers a counter.
+///
+/// Counters represent a single monotonic value, which means the value can only be incremented, not
+/// decremented, and always starts out with an initial value of zero.
+///
+/// Metrics can be registered, which provides a handle to directly update that metric.  For
+/// counters, [`Counter`](crate:Counter) is provided which can be incremented or set to an absolute value.
+///
+/// Metric names are shown below using string literals, but they can also be owned `String` values,
+/// which includes using macros such as `format!` directly at the callsite. String literals are
+/// preferred for performance where possible.
+///
+/// # Example
+/// ```
+/// # #![no_implicit_prelude]
+/// # use ::std::convert::From;
+/// # use ::std::format;
+/// # use ::std::string::String;
+/// # use metrics::register_counter;
+/// # fn main() {
+/// // A basic counter:
+/// let counter = register_counter!("some_metric_name");
+/// counter.increment(1);
+///
+/// // Specifying labels inline, including using constants for either the key or value:
+/// let counter = register_counter!("some_metric_name", "service" => "http");
+/// counter.absolute(42);
+///
+/// const SERVICE_LABEL: &'static str = "service";
+/// const SERVICE_HTTP: &'static str = "http";
+/// let counter = register_counter!("some_metric_name", SERVICE_LABEL => SERVICE_HTTP);
+/// counter.increment(123);
+///
+/// // We can also pass labels by giving a vector or slice of key/value pairs.  In this scenario,
+/// // a unit or description can still be passed in their respective positions:
+/// let dynamic_val = "woo";
+/// let labels = [("dynamic_key", format!("{}!", dynamic_val))];
+/// let counter = register_counter!("some_metric_name", &labels);
+///
+/// // As mentioned in the documentation, metric names also can be owned strings, including ones
+/// // generated at the callsite via things like `format!`:
+/// let name = String::from("some_owned_metric_name");
+/// let counter = register_counter!(name);
+///
+/// let counter = register_counter!(format!("{}_via_format", "name"));
+/// # }
+/// ```
+#[macro_export]
+macro_rules! register_counter {
+    (target: $target:expr, level: $level:expr, $name:expr $(, $label_key:expr $(=> $label_value:expr)?)* $(,)?) => {{
+        let metric_key = $crate::key_var!($name $(, $label_key $(=> $label_value)?)*);
+        let metadata = $crate::metadata_var!($target, $level);
+
+        $crate::recorder().register_counter(&metric_key, metadata)
+    }};
+    (target: $target:expr, $name:expr $(, $label_key:expr $(=> $label_value:expr)?)* $(,)?) => {
+        $crate::register_counter!(target: $target, level: $crate::Level::INFO, $name $(, $label_key $(=> $label_value)?)*)
+    };
+    (level: $level:expr, $name:expr $(, $label_key:expr $(=> $label_value:expr)?)* $(,)?) => {
+        $crate::register_counter!(target: module_path!(), level: $level, $name $(, $label_key$(=> $label_value)?)*)
```

Comment:
Should references to `module_path!()` also be qualified paths like you did above with `::std::vec!`?

Reply:
Oddly this `module_path` seems to still be imported when `no_implicit_prelude` is on, despite other macros like `println` not working.

As `module_path` can be shadowed also definitely should yeah.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=612b6e150612da97bd5d4dbc221ed40b

Reply:
11b36bc8c00c91d8b15db9a4bcda67ebe74d6549

Reply:
ah `no_implicit_prelude` does not disable https://doc.rust-lang.org/reference/names/preludes.html#language-prelude

Reply:
Makes sense. 👍🏻